<img src = "https://evangsmailoa.files.wordpress.com/2019/09/ml.png"  align = "center">

#<h1>Selamat datang di materi belajar Python!</h1>

<p><strong>Hoiilllaaa Hallo!</strong> Pada materi ini kita belajar tentang dasar-dasar main pemrograman dengan bahasa Python!</p>
<p>Selamat belajar!</p> 

#<h2 id="hello">Nulis "Hoiillaaaa Hallo dunia!" di Python</h2>

Biasa kan kalo mulai belajar bahasa pemrograman baru kita suka nulis "Hello World!", Nah ini sama seperti itu :)

<hr/>
<div class="alert alert-success alertsuccess" style="margin-top: 20px">
[Tip]: Untuk eksekusi, klik tanda **play** atau tekan <kbd>Shift</kbd> + <kbd>Enter</kbd> bersamaan.
</div>
<hr/>

In [ ]:
print('Kelas Machine Learning E!')

Kelas Machine Learning E!


In [ ]:
#ini tugas pertama
print ('Selamat datang Evangs Mailoa!')


Selamat datang Evangs Mailoa!


## <h3 id="version">Cari tahu versi Python yang kita gunakan?</h3>

<p>
    Saat ini ada 2 versi Python yang paling sering kita gunakan: Python 2 dan Python 3. 
</p>
<p>
    Nah untuk tahu Python versi berapa yang kita gunakan, langsung saja eksekusi koding berikut:
</p>

In [ ]:
# Cek versi Python

import sys
print(sys.version)

3.6.9 (default, Jul 17 2020, 12:50:27) 
[GCC 8.4.0]


## <h3 id="comments">Komentar di Python</h3>

<p>
    Biasanya kita kasi komentar di koding kita supaya besok2 kalo lupa bisa lihat komentar supaya ingat lagi. Nah kalo di Python kita gunakan simbol <code>#</code> sebelum komentar. Nanti waktu dieksekusi Python abaikan semua di belakang <code>#</code> jadi amaaaannn.
</p>

In [ ]:
#Latihan nulis komentar

print('Hello, Python!') #Baris ini tidak akan dieksekusi!
# print('Evangs Tetap Ganteng!')

Hello, Python!


## <h3 id="errors">Error yang sering muncul pas lagi main Python</h3>

<p>Everyone makes mistakes! Ya iyalah, tidak ada yang sempurna...Contoh, pas nulis <code>print</code> malah nulisnya <code>frint</code>, nanti Python akan bilang erros. Silakan dicoba!</p>

In [ ]:
#Coba pesan penanda error

print("Hello, Python!)

SyntaxError: ignored

In [ ]:
#Coba pesan penanda error

print("Hello, Python!")

## <h3 id="python_error">Python bisa tahu kode kita error atau gak sebelum dieksekusi?</h3>

Python itu kan <em>interpreted language</em> berbeda dengan <em> compiled languages</em>. Python akan hentikan eksekusi seluruh program kalo dia menemukan ada yang error.

In [ ]:
# Coba jalankan code berikut:

print("This will be printed")
frint("This will cause an error")
print("This will NOT be printed")

## <h3 id="exercise">LATIHAN: Tulislah program pertamamu bosku!</h3>

In [ ]:
# Tulis kodingmu yang isinya Hallo diikuti namamu! Kemudian tekan Shift+Enter untuk eksekusi.



# <h2 id="types_objects">Tipe data di Python</h2>

<p>Python itu kan termasuk OOP. Nah ada beberapa tipe objek di Python, antara lain: <i>strings</i>, <i>integers</i> dan <i>floats</i>. Setiap kali bro/siz tulis sesuatu (text) di Python, otomatis menggunakan <i>character strings</i> (strings for short). Kalo nulis angka berarti <i>integers</i> (contoh: -1, 0, 100) dan <i>floats</i>, yang mewakili angka real, atau kalo kite orang Indonesia suka sebuat angka2 koma2 (contoh: 3.14, -42.0).</p>

In [ ]:
# Ini integer

11

11

In [ ]:
# Ini float

2.14

2.14

In [ ]:
# Ini string

"Hello, Evangs Cakep!"

'Hello, Evangs Cakep!'

In [ ]:
# Cari tahu tipe dari 12 apa ya?

type(12)

str

In [ ]:
# Tipe dari 2.14?

type(2.14)

float

In [ ]:
# Tipe dari "Hello, Evangs Cakep!"

type("Hello, Python 101!")

str

## <h3 id="int">Integer</h3>

<p>Integer tuh bisa angka negatif maupun positif:</p>

<a align="center">
    <img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%201/Images/TypesInt.png" width="600">
</a>

In [ ]:
# Cetak tipe dari -1

type(-1)

int

In [ ]:
# Cetak tipe dari 4

type(4)

int

In [ ]:
# Cetak tipe dari 0

type(0)

int

## <h3 id="float">Float</h3> 

<p>Float tuh isinya angka real; kalo kita suka sebut angka koma2. Kalau mau detail float silakan cek <code>sys.float_info</code>. Nanti dikasi tahu seberapa besar dan kecil sebuah angka yang dapat diwakilkan oleh float.</p>

In [ ]:
# Cetak tipe dari 1.0

type(1.0) # 1 itu integer, tapi 1.0 itu float

float

In [ ]:
# Cetak tipe dari 0.5

type(0.5)

float

In [ ]:
# Cetak tipe dari 0.56

type(0.56)

float

In [ ]:
# System settings untuk tahu detail tipe float

sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

## <h3 id="convert">Konversi dari tipe satu ke tipe lain</h3>

<p>Di Python mudah untuk konversi; kalo di Java disebut overcasting dan ribet tapi kalo di Python gampang coy! Konversi dari tipe satu ke tipe lain disebut typecasting. Contoh, konversi dari <i>integer</i> ke <i>float</i> (contoh: 2 ke 2.0).</p>
<p>Markitcob! (Mari kita coba!)</p>

In [ ]:
# Pertama cari tahu tipe dari angka yang mau dikonversi, misalkan angka 2

type(2)

int

In [ ]:
# Mari kita konversi 2 ke float

float(2)

2.0

In [ ]:
# Nah sekarang sudah berubah itu bukan lagi int tapi sudah jadi float

type(float(2))

float

<p>Kalo konversi integer ke float, tidak terlalu berpengaruh terhadap nilainya. Namun, kalo kita cast angka float ke integer, pasti akan hilang beberapa informasi. Contoh, kalo cast angka float 1.1 ke integer maka yang kita dapat hanya angka 1 dan kita kehilangan informasi desimalnya (0.1):</p>

In [ ]:
# Casting 1.1 ke integer akan menyebabkan 'loss of information'

int(1.1)

1

<p>Kadang, kita punya string yang berisi angka2. nah untuk kasus begini, kita cast aja string itu ke integer pake <code>int()</code>:</p>

In [ ]:
# Konversi string ke integer

int('1')

1

<p>Tapi kalo kita coba string yang tidak sesuai dengan angka yang tepat, nanti malah error!</p> <p>Markitcob!:</p>

In [ ]:
# Konversi sebuah string ke sebuah angka integer

int('1 or 2 people')

ValueError: ignored

In [ ]:
# Konversi integer ke string

str(1)

'1'

In [ ]:
# Konversi float ke string

str(1.2)

'1.2'

## <h3 id="bool">Tipe Boolean</h3>

<p><i>Boolean</i> salah satu tipe yang penting juga di Python. Sebuah <i>Boolean</i> punya satu atau dua nilai: <code>True</code> or <code>False</code>:</p>

In [ ]:
# Bernilai TRUE

True

True

In [ ]:
# Bernilai FALSE

False

False

In [ ]:
type(True)

bool

In [ ]:
type(False)

bool

In [ ]:
# Konversi TRUE ke integer

int(True)

1

In [ ]:
# Konversi 1 ke boolean

bool(1)

True

In [ ]:
# Konversi 0 ke boolean

bool(0)

False

In [ ]:
# Konversi TRUE to float

float(True)

1.0

## <h3 id="exer_type">LATIHAN: </h3>

<p>Cari tahu tipe data dari: <code>6 / 2</code>?</p>

In [ ]:
# Tulis kodingmu kemudian eksekusi!


float

<p>Kira2 apa hasilnya: <code>6 // 2</code>? (Kalo pake <code>//</code>)</p>

# <h2 id="expressions">Ekspresi dan Variabel</h2>

<h3 id="exp">Ekspresi</h3>

<p>Ekspresi dalam Python berisi dasar operasi aritmatika seperti tambah, kurang, kali, bagi:</p>

In [ ]:
# Penjumlahan

43 + 60 + 16 + 41

160

In [ ]:
# Pengurangan

50 - 60

-10

In [ ]:
# Perkalian

5 * 5

25

In [ ]:
# Pembagian

25 / 6

4.166666666666667

In [ ]:
# Pembagian khusus mengambil integer

25 // 6

4

## <h3 id="exer_exp">LATIHAN: </h3>

<p>Tulis koding untuk menghitung berapa banyak jam untuk 160 menit?

In [ ]:
# Tulis kodingmu di bawah!



Klik 2X __disini__ untuk lihat jawaban.

<!-- Jawaban:
160/60 
# Atau
160//60
-->

In [ ]:
# Ekspresi Matika-matika

30 + 2 * 60

150

In [ ]:
# Ekspresi Matika-matika

(30 + 2) * 60

1920

## <h3 id="var">Variabel</h3>

In [ ]:
# Simpan hasil ke sebuah variabel X

X = 43 + 60 + 16 + 41

In [ ]:
# Cetak X

X

160

In [ ]:
# Kita bisa gunakan variabel tersebut untuk digabung ke variabel lainnya

Y = X / 60
Y

2.6666666666666665

## <h3 id="exer_exp_var">LATIHAN: </h3>

<p>Berapa nilai <code>X</code> jika <code> X = 3 + 2 * 2</code></p>

In [ ]:
# Tulis kodingmu kemudian eksekusi!

Klik 2X __disini__ untuk lihat jawaban.

<!-- Jawaban:
7
-->


<p>Berapa nilai <code>Y</code> jika <code>Y = (3 + 2) * 2</code>?</p>

In [ ]:
# Tulis kodingmu kemudian eksekusi!

Klik 2X __disini__ untuk lihat jawaban.

<!-- Jawaban:
10
-->

<p>Berapa nilai <code>Z</code> jika <code> Z = X + Y</code>?</p>

In [ ]:
# Tulis kodingmu kemudian eksekusi!

Klik 2X __disini__ untuk lihat jawaban.

<!-- Jawaban:
17
-->

# <h1>String dalam Python</h1>

# <h2 id="strings">Apa itu String?</h2>

Contoh berikut menampilkan sebuah string yang berisi 2 kata:

In [ ]:
# Contoh string yang dengan dua petik

"Evangs Mailoa"

'Evangs Mailoa'

In [ ]:
# Contoh string yang dengan satu petik

'Michael Jackson'

'Michael Jackson'

In [ ]:
# Contoh lain string

'@#2_#]&*^%$'

'@#2_#]&*^%$'

In [ ]:
# Contoh string disimpan dalam variabel Nama

Nama = "Michael Jackson"
Nama

'Michael Jackson'

# <h2 id="index">Indexing</h2>

Sebuah string tersusun rapi seperti array, yang disebut Index!   

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%201/Images/StringsIndex.png" width="600" align="center" />

Cara baca index-nya sebagai berikut:

In [ ]:
# Cetak elemen 1 dari string variabel Nama

print(Nama[0])

M


In [ ]:
# Cetak elemen ke-13

print(Nama[13])

o


## <h3 id="neg">Negative Indexing</h3>

 Di Python, kita juga bisa gunakan <b> negative indexing </b> dalam string:

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%201/Images/StringsNeg.png" width="600" align="center" />

Elemen terakhir diberikan index -1: 

In [ ]:
# Cetak elemen terakhir dalam variabel Nama

print(Nama[-15])

M


In [ ]:
# Panjangnya string

len("Michaelhsodfhsdifjdsif iosdifhsdiofhosdhfioihs Jackson")

54

## <h3 id="slice">Slicing</h3>

Kita dapat mengambil beberapa huruf dari string menggunakan <b>slicing</b>. Contoh kita ambil dari index 0 sampai 4, dan dari index 8 sampai 12:

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%201/Images/StringsSlice.png" width="600" align="center" />

In [ ]:
# Ambil slice (potongan) dari 0 sampai 3

Nama[0:4]

'Mich'

In [ ]:
# Ambil slice (potongan) dari 8 sampai 12

Nama[8:12]

'Jack'

## <h3 id="stride">Stride</h3>

<b>Stride</b> tuh ambil berdasarkan keputusan tertentu, misalkan kelipatan 2:

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%201/Images/StringsStride.png" width="600" align="center" />

In [ ]:
# Ambil semua elemen kedua. Elemen dalam index 1, 3, 5 dst.

Nama[::2]

'McalJcsn'

In [ ]:
# Atau ambil semua elemen kedua dari index 0 hingga index 4

Nama[0:7:5]

'Me'

## <h3 id="concat">Concatenate Strings</h3>

Maksudnya kita bisa potong-potong string dan menyambungnya sesuka hati:

In [ ]:
# Concatenate 2 string
Nama = "Evangs"
Statement = Nama + " memang terbaeeeeeek dah!"
Statement

In [ ]:
# Cetak string sebanyak 5X.

5 * "Evangs Ganteng"

# <h2 id="escape">Escape Sequences</h2>

Tanda "\n" berarti <b>new line</b> atau baris baru. Mirip-miriplah sama Bahasa C.

In [ ]:
#Buat baris baru

print("Evangs \n cakep banget")

Evangs 
 cakep banget


Tanda "\t" berarti <b>tab</b> atau menjorok ke dalam.

In [ ]:
#Buat tab

print("Evangs \t\t cakep banget")

Evangs 		 cakep banget


# <h2 id="operations">Operasi dalam String</h2>

Terdapat beberapa operasi string yang dapat kita lakukan dengan Python. Kita akan coba beberapa yang dasar ya.... Contoh: upper, lower, dsb. 

In [ ]:
# Konversi string ke huruf besar

A = "Teknik Informatika FTI UKSW Salatiga"
print("Sebelum konversi:", A)
B = A.upper()
print("Setelah konversi:", B)

Sebelum konversi: Teknik Informatika FTI UKSW Salatiga
Setelah konversi: TEKNIK INFORMATIKA FTI UKSW SALATIGA


In [ ]:
# Gantikan sub-string lama dengan sub-string baru.

A = "Evangs cakep banget"
B = A.replace('Evangs', 'Brad Pitt')
B

'Brad Pitt cakep banget'

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%201/Images/StringsFind.png" width="600" align="center" />

In [ ]:
# Cari index awal substring dari sebuah string. 

Name = "Michael Jackson"
Name.find('el')

5

In [ ]:
Name = "Michael Jackson"
Name.find('Jack')

8

# <h1>Tuples dalam Python</h1>

# <h2 id="tuple">Tuples</h2>

Dalam Python, terdapat tipe data: string, integer and float. Ketiganya dapat digabung dalam sebuah <b>tuple</b> sebagai contoh:

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/TuplesType.png" width="750" align="center" />

In [ ]:
# Buat sebuah tuple

tuple1 = ("disco",10,1.2 )
tuple1

('disco', 10, 1.2)

Tipe datanya adalah <b>tuple</b>

In [ ]:
# Cetak tipe 

type(tuple1)

tuple

## <h3 id="index">Indexing</h3>

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/TuplesIndex.gif" width="750" align="center">

Kita dapat mencetak masing-masing <b>tuple</b>

In [ ]:
# Cetak satu persatu isi tuple

print(tuple1[0])
print(tuple1[1])
print(tuple1[2])

disco
10
1.2


In [ ]:
# Cetak tipe dari masing-masing index

print(type(tuple1[0]))
print(type(tuple1[1]))
print(type(tuple1[2]))

<class 'str'>
<class 'int'>
<class 'float'>


KIta juga dapat menggunakan indexing negatif.

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/TuplesNeg.png" width="750" align="center">

In [ ]:
# Dengan indexing negatif untuk menampilkan elemen paling terakhir.

tuple1[-1]

1.2

In [ ]:
tuple1[-3]

'disco'

## <h3 id="concate">Concatenate Tuples</h3>

Kita dapat melakukan <b>concatenate</b> atau kombinasi <b>tuples</b> dengan menggunakan tanda **+** :

In [ ]:
# Concatenate dua tuple

tuple2 = tuple1 + ("hard rock", 10)
tuple2

('disco', 10, 1.2, 'hard rock', 10)

Kita juga dapat melakukan <b>slicing</b>.

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/TuplesSlice.gif" width="750" align="center">

In [ ]:
# Slice dari index 0 sampai index 2

tuple2[0:3]

('disco', 10, 1.2)

In [ ]:
# Slice dari index 3 sampai index 4

tuple2[3:5]

('hard rock', 10)

## <h3 id="sort">Sorting</h3>

Kita juga dapat melakukan <b>sorting</b> terhadap tuple.

In [ ]:
# Contoh sebuah tuple

Ratings = (0, 9, 6, 5, 10, 8, 9, 6, 2)

In [ ]:
# Sorting sebuah tuple

RatingsSorted = sorted(Ratings)
RatingsSorted

[0, 2, 5, 6, 6, 8, 9, 9, 10]

## <h3 id="nest">Nested Tuple</h3>

Bahasa kerennya tupel di dalam tuple.

In [ ]:
# Buat tuple di dalam tuple

NestedT =(1, 2, ("pop", "rock") ,(3,4),("disco",(1,2)))

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/TuplesNestOne.png" width="750" align="center">

In [ ]:
# Cetak setiap elemen berdasarkan index

print("Element 0 of Tuple: ", NestedT[0])
print("Element 1 of Tuple: ", NestedT[1])
print("Element 2 of Tuple: ", NestedT[2])
print("Element 3 of Tuple: ", NestedT[3])
print("Element 4 of Tuple: ", NestedT[4])

Element 0 of Tuple:  1
Element 1 of Tuple:  2
Element 2 of Tuple:  ('pop', 'rock')
Element 3 of Tuple:  (3, 4)
Element 4 of Tuple:  ('disco', (1, 2))


KIta juga dapat gunakan index kedua untuk akses ke tuple yang lain. Contoh:

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/TuplesNestTwo.png" width="750" align="center">

In [ ]:
# Cetak element dari masing-masing index, termasuk index tuple di dalamnya

print("Element 2, 0 of Tuple: ",   NestedT[2][0])
print("Element 2, 1 of Tuple: ",   NestedT[2][1])
print("Element 3, 0 of Tuple: ",   NestedT[3][0])
print("Element 3, 1 of Tuple: ",   NestedT[3][1])
print("Element 4, 0 of Tuple: ",   NestedT[4][0])
print("Element 4, 1 of Tuple: ",   NestedT[4][1])

In [ ]:
# Cetak elemen pertama di tuple kedua

NestedT[2][1][0]

'r'

# <h1>List dalam Python</h1>

# <h2 id="list">Lists</h2>

Untuk membuat <b>list</b>, jangan lupa taruh tanda <b>[ ]</b>, dan isi dalam list harus dipisah dengan koma. <p>Markitcob!</p>

In [ ]:
# Buat sebuah list

L = ["Michael Jackson", 10.1, 1982]
L

['Michael Jackson', 10.1, 1982]

Seperti biasa, kita dapat menggunakan indexing negatif.

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/ListsNeg.png" width="1000" />

In [ ]:
print('Elemen terakhir:' , L[-1] )

Elemen terakhir: 1982


## <h3 id="content">Isi List</h3>

Lists dapat berisi string, float, dan integer. Seperti biasa, kita bisa buat list dalam list. Contoh:

In [ ]:
# Contoh List

["Michael Jackson", 10.1, 1982, [1, 2], ("A", 1)]

['Michael Jackson', 10.1, 1982, [1, 2], ('A', 1)]

## <h3 id="op">Operasi List</h3>

Kita dapat melakukan slicing dalam list. Contoh, jika kita ingin ambil dua elemen terakhir maka kodingnya:

In [ ]:
# Buat sebuah List

L = ["Michael Jackson", 10.1,1982,"MJ",1]
L

['Michael Jackson', 10.1, 1982, 'MJ', 1]

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/ListsSlice.png" width="1000">

In [ ]:
# Lakukan slicing untuk list

L[3:5]

['MJ', 1]

Kita juga menambahkan isi list

In [ ]:
# Gunakan extend untuk menambahkan isi ke dalam list

L = [ "Michael Jackson", 10.2]
L.extend(['pop', 10])
L

Kita juga dapat menggunakan kode <b><code>append</code></b>:

In [ ]:
# Gunakan append untuk menambahkan isi ke dalam list

L = [ "Michael Jackson", 10.2]
L.append(['pop', 10])
L

List tuh <b>mutable</b>, sehingga kita dapat mengubah isinya suka-suka hati. sebagai contoh, kita ubah elemen pertamanya:



In [ ]:
# Ganti elemen berdasarkan index

A = ["disco", 10, 1.2]
print('Sebelum tukar:', A)
A[0] = 'hard rock'
print('Setelah tukar:', A)

Sebelum tukar: ['disco', 10, 1.2]
Setelah tukar: ['hard rock', 10, 1.2]


Kita juga dapat menghapus isi list, gunakan <b>del</b>:

In [ ]:
# Hapus elemen dengan del

print('Sebelum perubahan:', A)
del(A[0])
print('Setelah perubahan:', A)

Sebelum perubahan: ['hard rock', 10, 1.2]
Setelah perubahan: [10, 1.2]


Kita dapat konversi string ke dalam bentuk list menggunakan <b><code>split</code></b>:

In [ ]:
# Split sebuah string, default-nya adalah spasi

'hard rock'.split()

['hard', 'rock']

In [ ]:
# Split sebuah string dengan tanda koma

'A,B,C,D'.split(',')

['A', 'B', 'C', 'D']

## <h3 id="co">Copy dan Clone List</h3>

Saat kita seting sebuah variabel <b>B</b> sama dengan variabel <b>A</b>; keduanya <b>A</b> dan <b>B</b> merupakan list dalam memori:

In [ ]:
# Copy (copy by reference) isi list A

A = ["hard rock", 10, 1.2]
B = A
print('A:', A)
print('B:', B)

A: ['hard rock', 10, 1.2]
B: ['hard rock', 10, 1.2]


<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/ListsRef.png" width="1000" align="center">

In [ ]:
# Contoh lain

print('B[0]:', B[0])
A[0] = "banana"
print('B[0]:', B[0])

<img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/ListsRefGif.gif" width="1000" />

Kita bisa melakukan clone list  <b>A</b> dengan sintaks berikut:

In [ ]:
# Clone (clone by value) list A

B = A[:]
B

QUIZ# <h2 id="quiz">LATIHAN</h2>

Buat sebuah list <b><code>a_list</code></b>, yang isinya <b><code>1</code></b>,<b><code>hello</code></b>, <b><code>[1,2,3]</code></b> dan <b><code>True</code></b>. 

In [ ]:
# Silakan tulis koding di bawah dan tekan Shift+Enter untuk eksekusi

Tampilkan isi index 1 dari <b><code>a_list</code></b>.

In [ ]:
# Silakan tulis koding di bawah dan tekan Shift+Enter untuk eksekusi

Concatenate list tersebut dengan <code>A = [1, 'a']</code> and <code>B = [2, 1, 'd']</code>:

In [ ]:
# Silakan tulis koding di bawah dan tekan Shift+Enter untuk eksekusi

# <h1>Sets dalam Python</h1>

# <h2 id="set">Sets</h2>

Dalam Python ada kumpulan set. Nah tandanya pake kurung kurawal <b>{}</b>. Sets ini di dalamnya tidak boleh ada data yang sama atau double. Kalo ada maka secara otomatis akan dihapus oleh Python:

In [ ]:
# Buat set sederhana

set1 = {"pop", "rock", "soul", "hard rock", "rock", "R&B", "rock", "disco"}
set1

{'R&B', 'disco', 'hard rock', 'pop', 'rock', 'soul'}

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/SetsUnique.png" width="1100" />

## <h3 id="op">Set Operations</h3> 

In [ ]:
# Buat sebuah set

A = set(["Thriller", "Back in Black", "AC/DC"])
A

{'AC/DC', 'Back in Black', 'Thriller'}

Kita dapat menambah sebuah elemen ke dalam set menggunakan <b>add()</b>:

In [ ]:
# Tambah elemen ke dalam set

A.add("Didi Kempot")
A

{'AC/DC', 'Back in Black', 'Didi Kempot', 'Thriller'}

Kita dapat menghapus sebuah elemen dari dalam set menggunakan <b>remove()</b>:

In [ ]:
# Hapus elemen dari set

A.remove("Didi Kempot")
A

{'AC/DC', 'Back in Black', 'Thriller'}

Kita juga dapat melakukan cek apakah sebuah elemen ada dalam list:

In [ ]:
# Cek elemen dalam set

"AC/DC" in A

True

## <h3 id="logic">Operasi Logika dalam Set</h3>

Masih ingat materi dasar matika-matika? Masih ingat matika-matika diskrit? Nah mari kita lihat lebih jauh dalam Python!

In [ ]:
# Contoh buat dua buah set

album_set1 = set(["Thriller", 'AC/DC', 'Back in Black'])
album_set2 = set([ "AC/DC", "Back in Black", "The Dark Side of the Moon"])

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/SetsSamples.png" width="650" />

In [ ]:
# Cetak kedua set yang dibuat

album_set1, album_set2

({'AC/DC', 'Back in Black', 'Thriller'},
 {'AC/DC', 'Back in Black', 'The Dark Side of the Moon'})

Di kedua set sama-sama terdapat <b>AC/DC</b> dan <b>Back in Black</b> supaya mudah dipahami silakan lihat gambar berikut:

<img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/SetsLogic.png" width = "650" />

In [ ]:
# Cari intersection-nya (isi yang sama di kedua set)

intersection = album_set1 & album_set2
intersection

{'AC/DC', 'Back in Black'}

In [ ]:
# Cari data yang ADA di set1 tapi TIDAK ADA di set2

album_set1.difference(album_set2)  

{'Thriller'}

Perhatikan gambar berikut:

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/SetsLeft.png" width="650" />

Perhatikan gambar berikut:

<img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/SetsRight.png" width="650" />

Nah sekarang contoh mencari elemen yang HANYA ADA di <code>album_set2</code> dan TIDAK ADA di <code>album_set1</code>:

In [ ]:
album_set2.difference(album_set1)  

{'The Dark Side of the Moon'}

In [ ]:
# Gunakan method intersection untuk cek elemen yang ada baik di set album_list1 maupun di set album_list2

album_set1.intersection(album_set2)   

{'AC/DC', 'Back in Black'}

Gambarnya sebagai berikut:

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/SetsIntersect.png" width="650" />

Perhatikan gambar berikut yang menunjukkan semua data/elemen yang ada baik di set 1 maupun set 2:

<img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/SetsUnion.png" width="650" />

In [ ]:
# Cari union dari kedua set (data/elemen dalam semua set)

album_set1.union(album_set2)

{'AC/DC', 'Back in Black', 'The Dark Side of the Moon', 'Thriller'}

Nah kita dapat melakukan cek apakah set tersebut adalah <b>superset</b> atau <b>subset</b> dari set yang lain, seperti berikut:

In [ ]:
# Cek superset

set(album_set1).issuperset(album_set2)  

False

In [ ]:
# Cek subset

set(album_set2).issubset(album_set1)    

False

# <h1>Dictionary dalam Python</h1>

# <h2 id="Dic">Dictionaries</h2>

Sebuah <b>dictionary</b> terdiri dari kunci <b>(keys)</b> dan isi nilainya <b>(values)</b>. Nah untuk memudahkan bayangkan lah dictionary seperti list. Nah kalo list punya <b>index</b>, kalo dictionary punya <b>keys</b>.

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%202/Images/DictsList.png" width="650" />

In [ ]:
# Membuat sebuah dictionary

Dict = {"key1": 1, "key2": "2", "key3": [3, 3, 3], "key4": (4, 4, 4), ('key5'): 5, (0, 1): 6}
Dict

{(0, 1): 6,
 'key1': 1,
 'key2': '2',
 'key3': [3, 3, 3],
 'key4': (4, 4, 4),
 'key5': 5}

In [ ]:
# Ambil sebuah nilai dengan key-nya

Dict["key1"]

1

In [ ]:
# Contoh lain akses dengan key

Dict[(0, 1)]

6

In [ ]:
# Ambil semua isi dictionary

Dict.values() 

dict_values([1, '2', [3, 3, 3], (4, 4, 4), 5, 6])

In [ ]:
# Hapus beberapa isi dictionary dengan key-nya. Misalkan kita mau menghilangkan angka '5'

del(Dict['key5'])
Dict

{(0, 1): 6, 'key1': 1, 'key2': '2', 'key3': [3, 3, 3], 'key4': (4, 4, 4)}

# <h2 id="quiz">QUIZ-1</h2>

Diberikan sebuah tuple:

In [ ]:
# sample tuple

genres_tuple = ("pop", "rock", "soul", "hard rock", "soft rock", \
                "R&B", "progressive rock", "disco") 
genres_tuple

Carilah panjang dari tuple tersebut!

In [ ]:
# Tulis koding di bawah kemudian eksekusi atau tekan Shift+Enter

Akses elemen dalam index 3:

In [ ]:
# Tulis koding di bawah kemudian eksekusi atau tekan Shift+Enter

Gunakan slicing untuk mendapatkan index 3,4, dan 5!

In [ ]:
# Tulis koding di bawah kemudian eksekusi atau tekan Shift+Enter

Tampilkan dua elemen pertama dari tuple genre!

In [ ]:
# Tulis koding di bawah kemudian eksekusi atau tekan Shift+Enter

Tampilkan <b>"disco"</b>

In [ ]:
# Tulis koding di bawah kemudian eksekusi atau tekan Shift+Enter

Buat sorted list dari Tuple C_tuple=(-5, 1, -3):

In [ ]:
# Tulis koding di bawah kemudian eksekusi atau tekan Shift+Enter

Konversi list berikut <b><code>['rap','house','electronic music', 'rap']</code></b> ke dalam sebuah set:

In [ ]:
# Tuliskan kodingnya kemudian tekan Shift+Enter untuk eksekusi

Lakukan pengecekan misalkan set <b><code>A = [1, 2, 2, 1]</code></b> dan set <b><code>B = set([1, 2, 2, 1])</code></b>, apakah <b><code>sum(A) = sum(B)</code></b>

In [ ]:
# Tuliskan kodingnya kemudian tekan Shift+Enter untuk eksekusi

Buatlah sebuah set <b><code>album_set3</code></b> yang merupakan union dari <b><code>album_set1</code></b> dan <b><code>album_set2</code></b>:

In [ ]:
# Tuliskan kodingnya kemudian tekan Shift+Enter untuk eksekusi

album_set1 = set(["Thriller", 'AC/DC', 'Back in Black'])
album_set2 = set([ "AC/DC", "Back in Black", "The Dark Side of the Moon"])

Cek, apakah <b>album_set1</b> adalah subset dari <b>album_set3</b>:

In [ ]:
# Tuliskan kodingnya kemudian tekan Shift+Enter untuk eksekusi

<b>Buat dictionary berikut untuk menjawab dua soal di bawahnya:</b>

In [ ]:
# Buat sebuah dictionary

soundtrack_dic = {"The Bodyguard":"1992", "Saturday Night Fever":"1977"}
soundtrack_dic 

{'Saturday Night Fever': '1977', 'The Bodyguard': '1992'}

a) Dalam dictionary di atas, apa kuncinya <b>(key)</b> ?

In [ ]:
# Tuliskan kodingmu di bawah kemudian tekan Shift+Enter untuk eksekusi

b) Dalam dictionary <b><code>soundtrack_dict</code></b>, apa saja isinya <b>(value)</b>?

In [ ]:
# Tuliskan kodingmu di bawah kemudian tekan Shift+Enter untuk eksekusi

<b>DICTIONARY berikut untuk 4 soal di bawahnya:</b>
<p>
a) Buat dictionary yang berisi <b>Albums Back in Black, The Bodyguard </b>dan <b>Thriller</b> yang masing-masing mempunyai penjualan <b>50, 50 </b>dan <b>65 </b>juta keping.

In [ ]:
# Tuliskan kodingmu di bawah kemudian tekan Shift+Enter untuk eksekusi

b) Gunakan dictionary untuk mencari total penjualan dari <b>Thriller</b>:

In [ ]:
# Tuliskan kodingmu di bawah kemudian tekan Shift+Enter untuk eksekusi

c) Cari nama album dengan menggunakan <b><code>keys</code></b>:

In [ ]:
# Tuliskan kodingmu di bawah kemudian tekan Shift+Enter untuk eksekusi

d) cari hasil penjualan masing-masing album menggunakan <b><code>values</code></b>:

In [ ]:
# Tuliskan kodingmu di bawah kemudian tekan Shift+Enter untuk eksekusi

# <h2 id="(c)">(c)</h2>
<p>Copyright &copy; 2019 <b>Evangs Mailoa</b>.</p>

---


<p>Digunakan khusus untuk ngajar Machine Learning di Progdi Teknik Informatika - FTI UKSW</p>